### ISSUE TITLE
Is it possible to run cmaes in 'exploration' regime?

### ISSUE THREAD

**Issue by agrayver** [beniz/libcmaes#134]
Very often optimization methods are used to minimize misfit functions. Due to noise in data, all models below given noise level are essentially equally good. Obtaining an ensemble of such candidates (and not only one that gives minimal value) would allow a comprehensive characterization of an underlying model. 

I wonder, therefore, if CMAES (or any modifications of it) can be used to explore those low misfit regions of interest? One can of course try and run CMAES multiple times with random seed and starting guess, but it would be interesting to think if more intelligent methods can be derived?

**Comment by beniz** [comment]
This probably does not fully address your question, but have you tried the built-in uncertainty handling scheme ? If not, to use it, use `Parameters::set_uh(true)`.

The related paper should be:
Hansen, N., A.S.P. Niederberger, L. Guzzella and P. Koumoutsakos (2009). A Method for Handling Uncertainty in Evolutionary Optimization with an Application to Feedback Control of Combustion. IEEE Transactions on Evolutionary Computation, 13(1), pp. 180-197
see https://www.lri.fr/~hansen/publications.html

**Comment by agrayver** [comment]
Well, this is not quite what I was looking for, but I decided to play around with this option. Simply adding the following line cmaparams.set_uh(true); results in:

terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check

Here is call stack:
```
6	std::__throw_out_of_range(char const*)	/usr/lib/x86_64-linux-gnu/libstdc++.so.6		0x7ffff72eb5d7	
7	_M_range_check	stl_vector.h	794	0x7ffff798d4e5	
8	at	stl_vector.h	812	0x7ffff798d4e5	
9	libcmaes::CMASolutions::update_best_candidates	cmasolutions.cc	92	0x7ffff798d4e5	
10	libcmaes::CMAStrategy<libcmaes::ACovarianceUpdate, libcmaes::GenoPheno<libcmaes::NoBoundStrategy, libcmaes::NoScalingStrategy>>::tell	cmastrategy.cc	293	0x7ffff7955f16	
11	tell	cma_optimzer.h	69	0x4177b6	
12	CustomStrategy<libcmaes::CMAStrategy, libcmaes::ACovarianceUpdate, libcmaes::GenoPheno<libcmaes::NoBoundStrategy, libcmaes::NoScalingStrategy>>::optimize	cma_optimzer.h	103	0x4177b6	
13	optimize	esoptimizer.h	115	0x41afd4
```

**Comment by beniz** [comment]
Could you share the relevant part of your code please so that I can reproduce ? From the trace you have posted, I'd say you are overriding the main ask/tell loop and an initialization or pre-computation phase could be missing. Thanks.

**Comment by agrayver** [comment]
You're right, I'm using custom strategy where ask/tell simple call corresponding methods from the derived class.

To reproduce the error, you can take https://github.com/beniz/libcmaes/blob/master/examples/sample-code-ask-tell.cc and add cmaparams.set_uh(true);

**Comment by beniz** [comment]
Thanks, I see. When overriding `::eval()` the evaluation of the candidates required for uncertainty handling can be missing. Within the lib, this requires a bit of work as it relates to the overall availability of all options even when overriding the ask/tell functions, which may be difficult to enforce (as order may matter). I am not in a position where I can do this in the coming days.

So if you are still willing to test the uncertainty handling from within your custom code, and if you really cannot call on the original `eval` function, you will need to copy and add the conditional code that starts at https://github.com/beniz/libcmaes/blob/master/src/esostrategy.cc#L105

Let me know how it goes.

**Comment by agrayver** [comment]
Ok, I see. How about adding a method "perform_uh()" which would encapsulate logic starting at 
https://github.com/beniz/libcmaes/blob/master/src/esostrategy.cc#L105

and do nothing if _uh = false;

In this case it can simply be called from outside without knowing internal details. 
The same can be done with elitist mode which follows right after uh.

**Comment by beniz** [comment]
Yes of course, but there may be cases where order matters. I need to check that thoroughtly.
